File: nestery-backend/src/integrations/google-maps/google-maps.service.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import axios from 'axios';
import { LoggerService } from '../../core/logger/logger.service';
import { ExceptionService } from '../../core/exception/exception.service';

interface NearbyPlace {
  place_id: string;
  name: string;
  vicinity: string;
  types: string[];
  rating?: number;
  user_ratings_total?: number;
  geometry: {
    location: {
      lat: number;
      lng: number;
    };
  };
  photos?: Array<{
    photo_reference: string;
    height: number;
    width: number;
  }>;
  price_level?: number;
  opening_hours?: {
    open_now: boolean;
  };
}

interface PlacePhoto {
  photo_reference: string;
  height: number;
  width: number;
}

interface PlaceReview {
  author_name: string;
  rating: number;
  text: string;
  time: number;
}

interface PlacePrediction {
  place_id: string;
  description: string;
  structured_formatting: {
    main_text: string;
    secondary_text: string;
  };
  types: string[];
}

interface DirectionStep {
  distance: {
    text: string;
    value: number;
  };
  duration: {
    text: string;
    value: number;
  };
  html_instructions: string;
  travel_mode: string;
}

@Injectable()
export class GoogleMapsService {
  private readonly apiKey: string = '';
  private readonly mapsApiUrl: string = 'https://maps.googleapis.com/maps/api';
  private readonly routesApiUrl: string = 'https://routes.googleapis.com';

  constructor(
    private readonly configService: ConfigService,
    private readonly logger: LoggerService,
    private readonly exceptionService: ExceptionService,
  ) {
    const apiKey = this.configService.get<string>('GOOGLE_MAPS_API_KEY');
    if (apiKey) this.apiKey = apiKey;
    this.logger.setContext('GoogleMapsService');
  }

  /**
   * Geocode an address to get coordinates
   */
  async geocodeAddress(address: string) {
    try {
      this.logger.debug(`Geocoding address: ${address}`);

      const response = await axios.get(`${this.mapsApiUrl}/geocode/json`, {
        params: {
          address,
          key: this.apiKey,
        },
      });

      if (response.data.status !== 'OK') {
        throw new Error(`Geocoding failed: ${response.data.status}`);
      }

      const results = response.data.results;
      if (!results || results.length === 0) {
        throw new Error(`No results found for address: ${address}`);
      }

      const location = results[0].geometry.location;
      return {
        lat: location.lat,
        lng: location.lng,
        formattedAddress: results[0].formatted_address,
        placeId: results[0].place_id,
      };
    } catch (error) {
      this.logger.error(`Error geocoding address: ${error.message}`, error.stack);
      this.exceptionService.handleException(error);
      throw error;
    }
  }

  /**
   * Get nearby places based on location
   */
  async getNearbyPlaces(
    latitude: number,
    longitude: number,
    radius: number = 1000,
    type: string = 'restaurant',
  ) {
    try {
      this.logger.debug(
        `Getting nearby places at ${latitude},${longitude} with radius ${radius}m and type ${type}`,
      );

      const response = await axios.get(`${this.mapsApiUrl}/place/nearbysearch/json`, {
        params: {
          location: `${latitude},${longitude}`,
          radius,
          type,
          key: this.apiKey,
        },
      });

      if (response.data.status !== 'OK' && response.data.status !== 'ZERO_RESULTS') {
        throw new Error(`Nearby places search failed: ${response.data.status}`);
      }

      const places = response.data.results || [];

      return {
        success: true,
        places: places.map((place: NearbyPlace) => ({
          id: place.place_id,
          name: place.name,
          address: place.vicinity,
          types: place.types,
          rating: place.rating,
          reviewCount: place.user_ratings_total,
          latitude: place.geometry.location.lat,
          longitude: place.geometry.location.lng,
          photoReference: place.photos?.[0]?.photo_reference,
          priceLevel: place.price_level,
          openNow: place.opening_hours?.open_now,
        })),
      };
    } catch (error) {
      this.logger.error(`Error getting nearby places: ${error.message}`, error.stack);
      this.exceptionService.handleException(error);
      return {
        success: false,
        error: `Failed to get nearby places: ${error.message}`,
      };
    }
  }

  /**
   * Get place details
   */
  async getPlaceDetails(placeId: string) {
    try {
      this.logger.debug(`Getting place details for ID: ${placeId}`);

      const response = await axios.get(`${this.mapsApiUrl}/place/details/json`, {
        params: {
          place_id: placeId,
          fields:
            'name,formatted_address,geometry,rating,formatted_phone_number,website,opening_hours,photos,price_level,review',
          key: this.apiKey,
        },
      });

      if (response.data.status !== 'OK') {
        throw new Error(`Place details request failed: ${response.data.status}`);
      }

      const place = response.data.result;

      return {
        success: true,
        place: {
          id: place.place_id,
          name: place.name,
          address: place.formatted_address,
          latitude: place.geometry.location.lat,
          longitude: place.geometry.location.lng,
          rating: place.rating,
          phoneNumber: place.formatted_phone_number,
          website: place.website,
          openingHours: place.opening_hours?.weekday_text,
          openNow: place.opening_hours?.open_now,
          photos: place.photos?.map((photo: PlacePhoto) => ({
            reference: photo.photo_reference,
            width: photo.width,
            height: photo.height,
          })),
          priceLevel: place.price_level,
          reviews: place.reviews?.map((review: PlaceReview) => ({
            author: review.author_name,
            rating: review.rating,
            text: review.text,
            time: review.time,
          })),
        },
      };
    } catch (error) {
      this.logger.error(`Error getting place details: ${error.message}`, error.stack);
      this.exceptionService.handleException(error);
      return {
        success: false,
        error: `Failed to get place details: ${error.message}`,
      };
    }
  }

  /**
   * Get place autocomplete suggestions
   */
  async getPlaceAutocomplete(input: string, sessionToken: string, types: string = 'establishment') {
    try {
      this.logger.debug(`Getting place autocomplete for input: ${input}`);

      const response = await axios.get(`${this.mapsApiUrl}/place/autocomplete/json`, {
        params: {
          input,
          types,
          sessiontoken: sessionToken,
          key: this.apiKey,
        },
      });

      if (response.data.status !== 'OK' && response.data.status !== 'ZERO_RESULTS') {
        throw new Error(`Place autocomplete request failed: ${response.data.status}`);
      }

      const predictions = response.data.predictions || [];

      return {
        success: true,
        predictions: predictions.map((prediction: PlacePrediction) => ({
          id: prediction.place_id,
          description: prediction.description,
          mainText: prediction.structured_formatting.main_text,
          secondaryText: prediction.structured_formatting.secondary_text,
          types: prediction.types,
        })),
      };
    } catch (error) {
      this.logger.error(`Error getting place autocomplete: ${error.message}`, error.stack);
      this.exceptionService.handleException(error);
      return {
        success: false,
        error: `Failed to get place autocomplete: ${error.message}`,
      };
    }
  }

  /**
   * Get directions between two points
   */
  async getDirections(origin: string, destination: string, mode: string = 'driving') {
    try {
      this.logger.debug(`Getting directions from ${origin} to ${destination} via ${mode}`);

      const response = await axios.get(`${this.mapsApiUrl}/directions/json`, {
        params: {
          origin,
          destination,
          mode,
          key: this.apiKey,
        },
      });

      if (response.data.status !== 'OK') {
        throw new Error(`Directions request failed: ${response.data.status}`);
      }

      const routes = response.data.routes;
      if (!routes || routes.length === 0) {
        throw new Error('No routes found');
      }

      const route = routes[0];
      const leg = route.legs[0];

      return {
        success: true,
        route: {
          distance: leg.distance.text,
          distanceValue: leg.distance.value,
          duration: leg.duration.text,
          durationValue: leg.duration.value,
          startAddress: leg.start_address,
          endAddress: leg.end_address,
          startLocation: leg.start_location,
          endLocation: leg.end_location,
          steps: leg.steps.map((step: DirectionStep) => ({
            distance: step.distance.text,
            duration: step.duration.text,
            instructions: step.html_instructions,
            travelMode: step.travel_mode,
          })),
          polyline: route.overview_polyline.points,
        },
      };
    } catch (error) {
      this.logger.error(`Error getting directions: ${error.message}`, error.stack);
      this.exceptionService.handleException(error);
      return {
        success: false,
        error: `Failed to get directions: ${error.message}`,
      };
    }
  }
}

File: nestery-flutter/lib/services/google_maps_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:nestery_flutter/utils/constants.dart';
import 'package:nestery_flutter/utils/api_exception.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

class GoogleMapsService {
  final Dio _dio = Dio();
  final FlutterSecureStorage _secureStorage = const FlutterSecureStorage();
  
  // Singleton pattern
  static final GoogleMapsService _instance = GoogleMapsService._internal();
  
  factory GoogleMapsService() {
    return _instance;
  }
  
  GoogleMapsService._internal() {
    _initializeDio();
  }
  
  void _initializeDio() {
    _dio.options.baseUrl = 'https://maps.googleapis.com/maps/api';
    _dio.options.connectTimeout = const Duration(seconds: 10);
    _dio.options.receiveTimeout = const Duration(seconds: 10);
    _dio.options.headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };
    
    // Add interceptors for logging, error handling, etc.
    _dio.interceptors.add(InterceptorsWrapper(
      onRequest: (options, handler) async {
        // Add API key to all requests
        final apiKey = dotenv.env['GOOGLE_MAPS_API_KEY'] ?? AppConstants.googleMapsApiKey;
        options.queryParameters['key'] = apiKey;
        
        return handler.next(options);
      },
      onResponse: (response, handler) {
        return handler.next(response);
      },
      onError: (DioException e, handler) {
        return handler.next(e);
      },
    ));
  }
  
  // Geocode an address to get coordinates
  Future<Map<String, dynamic>> geocodeAddress(String address) async {
    try {
      final response = await _dio.get(
        '/geocode/json',
        queryParameters: {
          'address': address,
        },
      );
      
      if (response.data['status'] != 'OK') {
        throw ApiException(message: 'Geocoding failed: ${response.data['status']}');
      }
      
      final results = response.data['results'];
      if (results.isEmpty) {
        throw ApiException(message: 'No results found for address: $address');
      }
      
      final location = results[0]['geometry']['location'];
      return {
        'lat': location['lat'],
        'lng': location['lng'],
        'formatted_address': results[0]['formatted_address'],
      };
    } on DioException catch (e) {
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to geocode address: $e');
    }
  }
  
  // Reverse geocode coordinates to get address
  Future<Map<String, dynamic>> reverseGeocode(LatLng coordinates) async {
    try {
      final response = await _dio.get(
        '/geocode/json',
        queryParameters: {
          'latlng': '${coordinates.latitude},${coordinates.longitude}',
        },
      );
      
      if (response.data['status'] != 'OK') {
        throw ApiException(message: 'Reverse geocoding failed: ${response.data['status']}');
      }
      
      final results = response.data['results'];
      if (results.isEmpty) {
        throw ApiException(message: 'No results found for coordinates: $coordinates');
      }
      
      return {
        'formatted_address': results[0]['formatted_address'],
        'address_components': results[0]['address_components'],
      };
    } on DioException catch (e) {
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to reverse geocode: $e');
    }
  }
  
  // Get directions between two points
  Future<Map<String, dynamic>> getDirections({
    required LatLng origin,
    required LatLng destination,
    String mode = 'driving',
    List<LatLng>? waypoints,
    bool alternatives = false,
    String units = 'metric',
    String language = 'en',
  }) async {
    try {
      final response = await _dio.get(
        '/directions/json',
        queryParameters: {
          'origin': '${origin.latitude},${origin.longitude}',
          'destination': '${destination.latitude},${destination.longitude}',
          'mode': mode,
          'alternatives': alternatives.toString(),
          'units': units,
          'language': language,
          'waypoints': waypoints != null
              ? waypoints.map((point) => '${point.latitude},${point.longitude}').join('|')
              : null,
        },
      );
      
      if (response.data['status'] != 'OK') {
        throw ApiException(message: 'Directions request failed: ${response.data['status']}');
      }
      
      return response.data;
    } on DioException catch (e) {
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to get directions: $e');
    }
  }
  
  // Get places nearby
  Future<Map<String, dynamic>> getNearbyPlaces({
    required LatLng location,
    required double radius,
    String? type,
    String? keyword,
    String language = 'en',
  }) async {
    try {
      final response = await _dio.get(
        '/place/nearbysearch/json',
        queryParameters: {
          'location': '${location.latitude},${location.longitude}',
          'radius': radius.toString(),
          'type': type,
          'keyword': keyword,
          'language': language,
        },
      );
      
      if (response.data['status'] != 'OK' && response.data['status'] != 'ZERO_RESULTS') {
        throw ApiException(message: 'Nearby places request failed: ${response.data['status']}');
      }
      
      return response.data;
    } on DioException catch (e) {
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to get nearby places: $e');
    }
  }
  
  // Get place details
  Future<Map<String, dynamic>> getPlaceDetails({
    required String placeId,
    String language = 'en',
  }) async {
    try {
      final response = await _dio.get(
        '/place/details/json',
        queryParameters: {
          'place_id': placeId,
          'language': language,
          'fields': 'name,formatted_address,geometry,photos,rating,reviews,types,website,formatted_phone_number,opening_hours',
        },
      );
      
      if (response.data['status'] != 'OK') {
        throw ApiException(message: 'Place details request failed: ${response.data['status']}');
      }
      
      return response.data;
    } on DioException catch (e) {
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to get place details: $e');
    }
  }
  
  // Get distance matrix
  Future<Map<String, dynamic>> getDistanceMatrix({
    required List<LatLng> origins,
    required List<LatLng> destinations,
    String mode = 'driving',
    String units = 'metric',
    String language = 'en',
  }) async {
    try {
      final response = await _dio.get(
        '/distancematrix/json',
        queryParameters: {
          'origins': origins.map((point) => '${point.latitude},${point.longitude}').join('|'),
          'destinations': destinations.map((point) => '${point.latitude},${point.longitude}').join('|'),
          'mode': mode,
          'units': units,
          'language': language,
        },
      );
      
      if (response.data['status'] != 'OK') {
        throw ApiException(message: 'Distance matrix request failed: ${response.data['status']}');
      }
      
      return response.data;
    } on DioException catch (e) {
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to get distance matrix: $e');
    }
  }
  
  // Get autocomplete predictions
  Future<Map<String, dynamic>> getPlaceAutocomplete({
    required String input,
    String? sessionToken,
    LatLng? location,
    double? radius,
    String? types,
    String? components,
    bool strictbounds = false,
    String language = 'en',
  }) async {
    try {
      final response = await _dio.get(
        '/place/autocomplete/json',
        queryParameters: {
          'input': input,
          'sessiontoken': sessionToken,
          'location': location != null ? '${location.latitude},${location.longitude}' : null,
          'radius': radius?.toString(),
          'types': types,
          'components': components,
          'strictbounds': strictbounds.toString(),
          'language': language,
        },
      );
      
      if (response.data['status'] != 'OK' && response.data['status'] != 'ZERO_RESULTS') {
        throw ApiException(message: 'Place autocomplete request failed: ${response.data['status']}');
      }
      
      return response.data;
    } on DioException catch (e) {
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to get place autocomplete: $e');
    }
  }
  
  // Get static map image URL
  String getStaticMapUrl({
    required LatLng center,
    required int zoom,
    required int width,
    required int height,
    List<Map<String, dynamic>>? markers,
    List<Map<String, dynamic>>? paths,
    String mapType = 'roadmap',
    String format = 'png',
    String language = 'en',
  }) {
    final apiKey = dotenv.env['GOOGLE_MAPS_API_KEY'] ?? AppConstants.googleMapsApiKey;
    
    String url = 'https://maps.googleapis.com/maps/api/staticmap?';
    url += 'center=${center.latitude},${center.longitude}';
    url += '&zoom=$zoom';
    url += '&size=${width}x$height';
    url += '&maptype=$mapType';
    url += '&format=$format';
    url += '&language=$language';
    
    // Add markers
    if (markers != null && markers.isNotEmpty) {
      for (final marker in markers) {
        url += '&markers=';
        if (marker['color'] != null) url += 'color:${marker['color']}|';
        if (marker['label'] != null) url += 'label:${marker['label']}|';
        url += '${marker['lat']},${marker['lng']}';
      }
    }
    
    // Add paths
    if (paths != null && paths.isNotEmpty) {
      for (final path in paths) {
        url += '&path=';
        if (path['color'] != null) url += 'color:${path['color']}|';
        if (path['weight'] != null) url += 'weight:${path['weight']}|';
        
        final points = path['points'] as List<LatLng>;
        url += points.map((point) => '${point.latitude},${point.longitude}').join('|');
      }
    }
    
    url += '&key=$apiKey';
    
    return url;
  }
  
  // Get route using the new Routes API (replacing Directions API)
  Future<Map<String, dynamic>> getRoute({
    required LatLng origin,
    required LatLng destination,
    String travelMode = 'DRIVE',
    List<LatLng>? intermediates,
    String routingPreference = 'TRAFFIC_AWARE',
    String units = 'METRIC',
    String language = 'en',
  }) async {
    try {
      // Note: Routes API uses POST instead of GET
      final response = await _dio.post(
        '/routes/v2:computeRoutes',
        data: {
          'origin': {
            'location': {
              'latLng': {
                'latitude': origin.latitude,
                'longitude': origin.longitude,
              },
            },
          },
          'destination': {
            'location': {
              'latLng': {
                'latitude': destination.latitude,
                'longitude': destination.longitude,
              },
            },
          },
          'travelMode': travelMode,
          'routingPreference': routingPreference,
          'computeAlternativeRoutes': true,
          'routeModifiers': {
            'avoidTolls': false,
            'avoidHighways': false,
            'avoidFerries': false,
          },
          'languageCode': language,
          'units': units,
        },
      );
      
      return response.data;
    } on DioException catch (e) {
      throw ApiException.fromDioError(e);
    } catch (e) {
      throw ApiException(message: 'Failed to get route: $e');
    }
  }
}
